From: Tomasz Buchert <tomasz@debian.org>
Date: Sun, 18 Dec 2016 11:36:58 +0100
Subject: Add LDFLAGS to the final linking and sort $(wildcard) output

---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index c728037..8f347d6 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 OS := $(shell uname)
-LIBSOURCES = $(wildcard common/*.c) $(wildcard dec/*.c) $(wildcard enc/*.c)
+LIBSOURCES = $(sort $(wildcard common/*.c)) $(sort $(wildcard dec/*.c)) $(sort $(wildcard enc/*.c))
 SOURCES = $(LIBSOURCES) tools/bro.c
 BINDIR = bin
 OBJDIR = $(BINDIR)/obj
@@ -27,7 +27,7 @@ $(OBJECTS): $(DIRS)
         -c $(patsubst %.o,%.c,$(patsubst $(OBJDIR)/%,%,$@)) -o $@
 
 $(EXECUTABLE): $(OBJECTS)
-	$(CC) $(OBJECTS) -lm -o $(BINDIR)/$(EXECUTABLE)
+	$(CC) $(LDFLAGS) $(OBJECTS) -lm -o $(BINDIR)/$(EXECUTABLE)
 
 lib: $(LIBOBJECTS)
 	rm -f $(LIB_A)
