Description: Fix integer underflow
 Fixes for CVE-2016-1968 and CVE-2016-1624
Origin: https://github.com/google/brotli/commit/37a320dd81db8d546cd24a45b4c61d87b45dcade
Bug-Debian: https://bugs.debian.org/817233
Index: brotli-0.3.0+dfsg/dec/decode.c
===================================================================
--- brotli-0.3.0+dfsg.orig/dec/decode.c	2015-12-22 08:28:21.000000000 -0800
+++ brotli-0.3.0+dfsg/dec/decode.c	2016-03-14 08:46:10.635050201 -0700
@@ -1688,6 +1688,10 @@
   } else {
     const uint8_t *ringbuffer_end_minus_copy_length =
         s->ringbuffer_end - i;
+    /* Check for possible underflow and clamp the pointer to 0. */
+    if (PREDICT_FALSE(s->ringbuffer_end < (const uint8_t*)0 + i)) {
+        ringbuffer_end_minus_copy_length = 0;
+    }
     uint8_t* copy_src = &s->ringbuffer[
         (pos - s->distance_code) & s->ringbuffer_mask];
     uint8_t* copy_dst = &s->ringbuffer[pos];
